cmd/clusterctl/client/cluster/topology.go

```diff
@@ -0,0 +1,613 @@
+		return errors.Wrap(err, "failed to set missing namespaces")
+	}
+
+	if err := t.prepareClusters(getClusters(in.Objs), apiReader); err != nil {
+		return errors.Wrap(err, "failed to prepare clusters")
+	}
+	return nil
+}
+
+// setMissingNamespaces sets the object to the current namespace on objects
+// that are missing the namespace field.
+func (t *topologyClient) setMissingNamespaces(in *DryRunInput) error {
+	currentNamespace, err := t.proxy.CurrentNamespace()
+	if err != nil {
+		return errors.Wrap(err, "failed to get current namespace")
+	}
+	for i := range in.Objs {
+		if in.Objs[i].GetNamespace() == "" {
+			in.Objs[i].SetNamespace(currentNamespace)
+		}
+	}
+	return nil
+}
+
+// prepareClusters does the following operations on each Cluster in the input.
+// - Check if the Cluster exists in the real apiserver.
+// - If the Cluster exists in the real apiserver we merge the object from the
+//   server with the object from the input. This final object correctly represents the
+//   modified cluster object.
+// *Important note*: We do this above operation because the topology reconciler in a
+//   real run takes as input a cluster object from the apiserver that has merged spec of
+//   the changes in the input and the one stroed in the server. For example: the cluster
+//   object in the input will not have cluster.spec.infrastructureRef and cluster.spec.controlPlaneRef
+//   but the merged object will have these fields set.
+func (t *topologyClient) prepareClusters(clusters []*unstructured.Unstructured, apiReader client.Reader) error {
+	if apiReader == nil {
+		// If there is no backing server there is nothing more to do here.
+		// Return early.
+		return nil
+	}
+
+	ctx := context.TODO()
+	// For a Cluster check if it already exists in the server. If it does, get the object from the server
+	// and merge it with the Cluster from the file to get effective 'modified' Cluster.
+	for _, cluster := range clusters {
+		storedCluster := &clusterv1.Cluster{}
+		if err := apiReader.Get(
+			ctx,
+			client.ObjectKey{Namespace: cluster.GetNamespace(), Name: cluster.GetName()},
+			storedCluster,
+		); err != nil {
+			if apierrors.IsNotFound(err) {
+				// The Cluster does not exist in the server. Nothing more to do here.
+				continue
+			}
+			return errors.Wrapf(err, "failed to get Cluster %s/%s", cluster.GetNamespace(), cluster.GetName())
+		}
+		originalJSON, err := json.Marshal(storedCluster)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert Cluster %s/%s to json", storedCluster.Namespace, storedCluster.Name)
+		}
+		modifiedJSON, err := json.Marshal(cluster)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert Cluster %s/%s", cluster.GetNamespace(), cluster.GetName())
+		}
+		// Apply the modified object to the original one, merging the values of both;
+		// in case of conflicts, values from the modified object are preserved.
+		originalWithModifiedJSON, err := jsonpatch.MergePatch(originalJSON, modifiedJSON)
```

Comment:
I'm not too familiar with what kubectl is doing per default. Is that the same? Should we add a disclaimer in the cmd doc that we're doing it like this to keep it simple? (I think it's a simplification given the different possible client and server-side apply/merge methods available)

Reply:
Absolutely agree that we should keep our merge logic simple